File: ReverseLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional


class Solution:
    # def reverseList2(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     if head is None: return None
    #     retNode = ListNode(head.val)
    #     while head.next is not None:
    #         head = head.next
    #         retNode = ListNode(head.val, retNode)
    #     return retNode 
    
    # in-place
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]: # type: ignore
        prevNode = None
        while head:
            nextNode = head.next
            head.next = prevNode
            prevNode = head
            head = nextNode
        return prevNode    
        
        

File: ReverseLinkedList.java
public class ReverseLinkedList {
    public class ListNode {
            int val;
            ListNode next;
            ListNode() {}
            ListNode(int val) { this.val = val; }
            ListNode(int val, ListNode next) { this.val = val; this.next = next; }
        }
    public ListNode reverseList(ListNode h) {
        if (h == null) return null;
        // print(h);
        ListNode head = h;
        ListNode ret = new ListNode(head.val);
        while (head.next != null) {
            head = head.next;
            ret = new ListNode(head.val, ret);
            // print(ret);
        }
        return ret;
    }
    public void print(ListNode n) {
        if (n == null) return;
        ListNode t = n;
        String str = "";
        while (t.next != null) {
            str += t.val;
            t = t.next;
        }
        str += t.val;
        System.out.println(str);
    }
}
